ppac-service-impl/src/main/kotlin/com/levi9/ppac/service/api/service/CompanyServiceImpl.kt

```diff
@@ -17,16 +20,68 @@ import java.util.*
 class CompanyServiceImpl(
     private val securityContext: SecurityContext<Int>,
     private val companyRepository: CompanyRepository,
-    private val codeRepository: CodeRepository
+    private val codeRepository: CodeRepository,
+    private val openingRepository: OpeningRepository,
+    private val companyCodeRepository: CompanyCodeRepository
 ) : CompanyService<Company> {
+
+    @Transactional
+    override fun addOpening(id: UUID, opening: Opening): Opening {
+
+        if (!codeRepository.isCompanyCode(securityContext.getAccessCode(), id)) {
+            throw ResponseStatusException(HttpStatus.UNAUTHORIZED)
+        }
+
+        val companySet = opening.trainers.map { companyRepository.findCompanyEntitiesByOpenings_Trainers_Id(it.id) }
+            .flatten()
+            .toSet()
+
+        if (companySet.isNotEmpty() && (companySet.size > 1 || companySet.first().id != id))
+            throw ResponseStatusException(HttpStatus.BAD_REQUEST)
+
+        val openingDTO = opening.apply { this.id = UUID.randomUUID() }
+        val openingEntity = Opening.parse(openingDTO)
+        val savedOpening = openingRepository.save(openingEntity)
+
+        val companyEntity = companyRepository.findByIdOrNull(id)!!
+        companyEntity.openings += savedOpening
+        companyRepository.save(companyEntity)
+
+        return Opening.parse(savedOpening)
+    }
+
     @Transactional
     override fun findAll(): List<Company> {
+        return companyRepository.findAll().map { companyEntity ->
+            val activeOpenings = companyEntity.openings?.filter { it.available }
+            companyEntity.openings = activeOpenings
+            Company.parse(companyEntity)
+        }
+    }
 
-        if (!codeRepository.isAdminCode(securityContext.getAccessCode())) {
+    @Transactional
+    override fun findById(id: UUID): Company {
+        val company = companyRepository.findById(id).orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND) }
+        val activeOpenings = company.openings?.filter { it.available }
```

Comment:
The sign '?' it is not necessary. A warn "Unnecessary safe call on a non-null receiver of type List<OpeningEntity>" is present.  You can resolve this by removing '?' . Do this for all warn cases in this class.